filebeat/registrar/registrar.go

```diff
@@ -183,24 +185,51 @@ func (r *Registrar) Run() {
 // onEvents processes events received from the publisher pipeline
 func (r *Registrar) onEvents(states []file.State) {
 	r.processEventStates(states)
+	r.bufferedStateUpdates += len(states)
+
+	// check if we need to enable state cleanup
+	if !r.gcEnabled {
+		for i := range states {
+			if states[i].TTL >= 0 || states[i].Finished {
+				r.gcEnabled = true
+				break
+			}
+		}
+	}
+
+	logp.Debug("registrar", "Registrar state updates processed. Count: %v", len(states))
+
+	// new set of events received -> mark state registry ready for next
+	// cleanup phase in case gc'able events are stored in the registry.
+	r.gcRequired = r.gcEnabled
+}
+
+// gcStates runs a registry Cleanup. The bool returned indicates wether more
+// events in the registry can be gc'ed in the future.
+func (r *Registrar) gcStates() {
+	if !r.gcRequired {
+		return
+	}
 
 	beforeCount := r.states.Count()
-	cleanedStates := r.states.Cleanup()
+	cleanedStates, pendingClean := r.states.Cleanup()
 	statesCleanup.Add(int64(cleanedStates))
 
-	r.bufferedStateUpdates += len(states)
-
 	logp.Debug("registrar",
-		"Registrar states cleaned up. Before: %d, After: %d",
-		beforeCount, beforeCount-cleanedStates)
+		"Registrar states cleaned up. Before: %d, After: %d, Pending: %d",
+		beforeCount, beforeCount-cleanedStates, pendingClean)
+
+	r.gcRequired = false
+	r.gcEnabled = pendingClean > 0
```

Comment:
@urso Looking at this PR I now better understand why you introduced `pendingClean`. I think the confusing part for me here is the name of the variable. pending indicates for me it's already in a queue or similar for cleanup but has not finished yet. But what we have here is states which have an expiry time stamp but are not expired yet. Unexpired is the word that comes to my mind but is not nice.

It would be nice to know which states were removed because of `clean_inactive` and which ones because of `clean_removed`. So fare we don't make a difference. It would me we would log 4 stats:

* Total states after cleanup without expiry 
* Total states after cleanup with expiry 
* States removed because of expiry
* States removed because of removal

Based on this the number of previous states could be calculated.

Reply:
I think theses addition makes sense, it's mostly for us for debugging registry problems.

Reply:
Let's do the stats reporting in another PR. Here I'd like to focus on deferring the cleanup